src/ripple/app/ledger/impl/LedgerCleaner.cpp

```diff
@@ -96,20 +101,24 @@ class LedgerCleanerImp
     //
     //--------------------------------------------------------------------------
 
-    void onPrepare ()
+    void onPrepare () override
     {
     }
 
-    void onStart ()
+    void onStart () override
     {
-        startThread();
+        thread_ = std::thread {&LedgerCleanerImp::run, this};
     }
 
-    void onStop ()
+    void onStop () override
     {
-        m_journal.info << "Stopping";
-        signalThreadShouldExit();
-        notify();
+        JLOG (j_.info) << "Stopping";
+        {
+            std::lock_guard<std::mutex> lock (mutex_);
+            state_ = State::shouldExit;
+            wakeup_.notify_one();
```

Comment:
No deadlock, this is ok.  It is true that the signaled thread won't exit the wait until this thread gives up the mutex.  But that thread will be signaled, blocked on the mutex, and will continue once it obtains the lock on the mutex.